File: python/mergeIntervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def merge(self, intervals: List[Interval]) -> List[Interval]:
        if not intervals:
            return []
        intervals = sorted(intervals, key=lambda x: (x.start, x.end))
        m_intervals = [Interval(intervals[0].start, intervals[0].end)]
        for c_int in intervals[1:]:
            if c_int.end > m_intervals[-1].end:
                if c_int.start <= m_intervals[-1].end:
                    m_intervals[-1].end = c_int.end
                else:
                    m_intervals.append(c_int)
        return m_intervals

File: java/mergeIntervals.java
class Interval {
    public int start;
    public int end;
    public Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
}

class Solution {
    public int[][] merge(int[][] intervalArray) {
        ArrayList<Interval> intervals = new ArrayList<>();
        Arrays.sort(intervalArray, new Comparator<int[]>(){
            @Override
            public int compare(int[] a, int[] b){
                return a[0] - b[0];
            }
        });
        
        for (int[] interval : intervalArray) {
            if (!intervals.isEmpty() && 
                interval[0] <= intervals.get(intervals.size() - 1).end) {
                if (interval[1] > intervals.get(intervals.size() - 1).end)
                    intervals.get(intervals.size() - 1).end = interval[1];
            }
            else {
                intervals.add(new Interval(interval[0], interval[1]));
            }
        }
        
        int[][] returnArray = new int[intervals.size()][2];
        
        for (int i0 = 0; i0 < intervals.size(); i0++) {
            returnArray[i0][0] = intervals.get(i0).start;
            returnArray[i0][1] = intervals.get(i0).end;
        }
        
        return returnArray;
    }
}
